In [95]:
import numpy as np
import pandas as pd
# https://plotly.com/python/ 
# good for interactive graphs
import plotly.express as px

# Project - Data Cleaning 



In [96]:
df = pd.read_csv('https://raw.githubusercontent.com/ReDI-School/data-analytics/master/Class%20material/Class-11_Data_cleaning/berlin_housing_with_scraped.csv')

## Check some basic info

In [97]:
df.shape

(764, 22)

In [98]:
df.head(10)

url  ...                                               tags
0  https://www.immobilienscout24.de/expose/116051687  ...  ['Balkon', 'Terrasse Balkon', 'Terrasse Person...
1  https://www.immobilienscout24.de/expose/115338103  ...  ['Balkon', 'Terrasse Balkon', 'Terrasse Person...
2  https://www.immobilienscout24.de/expose/116458710  ...  ['Balkon', 'Terrasse Balkon', 'Terrasse Person...
3  https://www.immobilienscout24.de/expose/116573177  ...  ['Balkon', 'Terrasse Balkon', 'Terrasse Person...
4  https://www.immobilienscout24.de/expose/115925878  ...  ['Balkon', 'Terrasse Balkon', 'Terrasse Person...
5  https://www.immobilienscout24.de/expose/115611847  ...  ['Balkon', 'Terrasse Balkon', 'Terrasse Person...
6  https://www.immobilienscout24.de/expose/108376992  ...  ['Balkon', 'Terrasse Balkon', 'Terrasse Person...
7  https://www.immobilienscout24.de/expose/116573270  ...  ['Balkon', 'Terrasse Balkon', 'Terrasse Person...
8  https://www.immobilienscout24.de/expose/116456427  ...  ['Balkon', 'Terrasse Balkon', 'Terrasse Person...
9  https://www.immobilienscout24.de/expose/113934099  ...  ['Balkon', 'Terrasse Balkon', 'Terrasse Person...

[10 rows x 22 columns]

In [99]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 764 entries, 0 to 763
Data columns (total 22 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   url                  764 non-null    object 
 1   region               764 non-null    object 
 2   condition            764 non-null    object 
 3   rooms                764 non-null    float64
 4   rent                 764 non-null    float64
 5   year_construction    764 non-null    int64  
 6   space                764 non-null    float64
 7   cold_rent            426 non-null    object 
 8   utilities            426 non-null    object 
 9   heating_costs        426 non-null    object 
 10  security_deposit     385 non-null    object 
 11  parking_space        90 non-null     object 
 12  renovated_date       195 non-null    float64
 13  property_condition   501 non-null    object 
 14  furnishing           445 non-null    object 
 15  heating_type         515 non-null    obj

In [100]:
(df.isna().sum() / len(df)).sort_values(ascending = False)

parking_space          0.882199
renovated_date         0.744764
efficiency_class       0.671466
energy_requirement     0.573298
security_deposit       0.496073
heating_costs          0.442408
cold_rent              0.442408
utilities              0.442408
furnishing             0.417539
energy_certificate     0.409686
type_of_certificate    0.388743
energy_sources         0.353403
property_condition     0.344241
heating_type           0.325916
rooms                  0.000000
region                 0.000000
condition              0.000000
tags                   0.000000
rent                   0.000000
year_construction      0.000000
space                  0.000000
url                    0.000000
dtype: float64

### Clean `tags`

Have a look at the tags columns. What does this information tell us? Notice that the column doesn't contain useful information. In fact, it's just the same value repeated. So we can just remove that column completely from the dataset.

In [101]:
df['tags'][0]

"['Balkon', 'Terrasse Balkon', 'Terrasse Personenaufzug Personenaufzug Einbauküche Einbauküche']"

In [102]:
df['tags'].unique()

array(["['Balkon', 'Terrasse Balkon', 'Terrasse Personenaufzug Personenaufzug Einbauküche Einbauküche']"],
      dtype=object)

In [103]:

df.nunique().sort_values()

tags                     1
energy_certificate       3
heating_type             7
efficiency_class         8
furnishing               8
condition                9
property_condition       9
region                  10
rooms                   11
renovated_date          12
parking_space           21
energy_sources          29
type_of_certificate     85
year_construction       93
heating_costs          111
energy_requirement     113
security_deposit       216
utilities              261
cold_rent              354
space                  571
rent                   639
url                    764
dtype: int64

In [104]:
df.drop('tags', axis=1, inplace=True)

In [105]:
df.shape

(764, 21)

## Clean `security_deposit`



In [106]:
df['deposit'] = df['security_deposit']

In [107]:
df['deposit'].value_counts()

3           123
Drei         13
Kaution:      7
3-fache       5
4350          3
           ... 
4800          1
2.085,00      1
2315          1
4759          1
3.750,00      1
Name: deposit, Length: 216, dtype: int64

In [108]:
df['deposit'].unique()

array(['3', '2920,53', 'Drei', nan, '2400', '3507.42', '2000', '2139.78',
       '5.397,00', '3885,00', '7.950,00', '4152', '4.650,00', '1.500€',
       '3x', '9.150,-', '4.545', '2xNKM', '6.900,00', '5.994,00',
       '6.000,-', '7.470,00', '5.172,00', '2.145,00', '6.215,70',
       '11.475,00', '14.850,00', '4.197,00', '2600,00', '12.900,00', '2',
       '3450', '5000', '10.000', '3620', '1690', '1790', 'ja', '2990',
       '1490', '9.960,00', '3159,00', '4.350,00', '5010', '4333,90',
       '3000', '2.013,00', '2841', '4000', '2850.72', '3300.75', '2344',
       '3645,00', '2397', '4500', '9.718,00', '3.597,00', '3900',
       '3.780,00', '4800', '3.582', '3630', '2549.79', '3831', '2.274,00',
       '2850', '2960', '4.680', '4350', '3500', '4080', '4050', '2338',
       '5940,00', '8100', '3.859,17', '2.630,43', '6552', '2.700,00',
       '1797,00', '3345', '3360', '5148', '2472', '12.000,00', '6.300,00',
       '5.394,00€', '1.554,00', '6600€', '1.727,76', '9025,50',
       '3.750

### Remove € 

In [109]:
df['deposit'][df['deposit'].str.contains("€", na=False)] # Series.contains("€")

36        1.500€
167       1.500€
271    5.394,00€
289        6600€
393       3.597€
403       3.750€
405        5097€
423    2.286,45€
462      6570.0€
464      6675.0€
466      9333.0€
469      9792.0€
490     1074,60€
507    4.698,00€
545    4.197,00€
697        €3500
Name: deposit, dtype: object

In [110]:
df['deposit'] = df['deposit'].apply(lambda deposit: str(deposit).replace('€', ''))

In [111]:
df['deposit'].unique()

array(['3', '2920,53', 'Drei', 'nan', '2400', '3507.42', '2000',
       '2139.78', '5.397,00', '3885,00', '7.950,00', '4152', '4.650,00',
       '1.500', '3x', '9.150,-', '4.545', '2xNKM', '6.900,00', '5.994,00',
       '6.000,-', '7.470,00', '5.172,00', '2.145,00', '6.215,70',
       '11.475,00', '14.850,00', '4.197,00', '2600,00', '12.900,00', '2',
       '3450', '5000', '10.000', '3620', '1690', '1790', 'ja', '2990',
       '1490', '9.960,00', '3159,00', '4.350,00', '5010', '4333,90',
       '3000', '2.013,00', '2841', '4000', '2850.72', '3300.75', '2344',
       '3645,00', '2397', '4500', '9.718,00', '3.597,00', '3900',
       '3.780,00', '4800', '3.582', '3630', '2549.79', '3831', '2.274,00',
       '2850', '2960', '4.680', '4350', '3500', '4080', '4050', '2338',
       '5940,00', '8100', '3.859,17', '2.630,43', '6552', '2.700,00',
       '1797,00', '3345', '3360', '5148', '2472', '12.000,00', '6.300,00',
       '5.394,00', '1.554,00', '6600', '1.727,76', '9025,50', '3.750,00',
  

### Convert values that "look" like numbers to parseable formats

The deposit column is of type string, but most values are clearly indicating some concrete numeric value (we ignore values like `3, drei, etc.` for now). However, there is a mix of formats which makes it difficult to convert the amounts to actual numeric values.

Note that the german number format has different separators for the thousand places and decimal places. Concretely, `.` separates the thousand places and `,` is for the decimals. `1.200,50` in the german format corresponds to `1,200.50` in the US format.

Looking at the dataset, we have these different formats

* 2920,53 --> german number format, only decimal separator (cannot be converted directly to float)
* 4.650,00 --> german number format with thousand seperator and decimal separator (cannot be converted directly to float)
* 2400 --> plain format (can be converted directly to float) 
* 3507.42 --> US number format (can be converted directly to float)
* 10.000 --> german number format, only thousand decimal separator (converting to float would work but would be interpreted as 10, not 10-thousand)



In [112]:
def convert_string_to_numeric_representation(value):
    """Replace a period with empty stringm and comma with period (decimal point)."""
    return value.replace(".", "").replace(",", ".")


def convert_decimal(value):
    """Convert numeric string to a decimal number."""
    
    # return NaN if missing
    if pd.isna(value):
        return np.nan       
    
    # check if wrong format without decimal i.e. 3.500
    if len(value.split('.')[-1]) == 3:
        return convert_string_to_numeric_representation(value)
    
    # convert to float
    try:                               
        return float(value)
    except:
        return convert_string_to_numeric_representation(value)

In [113]:
df['deposit'] = df['deposit'].apply(convert_decimal)
df['deposit'].unique()

array([3.0, '2920.53', 'Drei', 'nan', 2400.0, 3507.42, 2000.0, 2139.78,
       '5397.00', '3885.00', '7950.00', 4152.0, '4650.00', '1500', '3x',
       '9150.-', '4545', '2xNKM', '6900.00', '5994.00', '6000.-',
       '7470.00', '5172.00', '2145.00', '6215.70', '11475.00', '14850.00',
       '4197.00', '2600.00', '12900.00', 2.0, 3450.0, 5000.0, '10000',
       3620.0, 1690.0, 1790.0, 'ja', 2990.0, 1490.0, '9960.00', '3159.00',
       '4350.00', 5010.0, '4333.90', 3000.0, '2013.00', 2841.0, 4000.0,
       2850.72, 3300.75, 2344.0, '3645.00', 2397.0, 4500.0, '9718.00',
       '3597.00', 3900.0, '3780.00', 4800.0, '3582', 3630.0, 2549.79,
       3831.0, '2274.00', 2850.0, 2960.0, '4680', 4350.0, 3500.0, 4080.0,
       4050.0, 2338.0, '5940.00', 8100.0, '3859.17', '2630.43', 6552.0,
       '2700.00', '1797.00', 3345.0, 3360.0, 5148.0, 2472.0, '12000.00',
       '6300.00', '5394.00', '1554.00', 6600.0, '1727.76', '9025.50',
       '3750.00', 7575.0, '7650.00', '5355.00', 5850.0, '4170.00',

### Map values that cannot be interpreted as numeric

In [114]:
def is_number(value):
  try:
    float(value)
    return True
  except:
    return False 

# different alternatives for negating a boolean check
#print(~df['deposit'].apply(is_number))
#print(df['deposit'].apply(lambda value: not is_number(value)))

df['deposit'][~df['deposit'].apply(is_number)].unique()

array(['Drei', '3x', '9150.-', '2xNKM', '6000.-', 'ja', '3-fache',
       'Kaution:', 'Kaution', '3NKM'], dtype=object)

Build a map of what the values should be. This is manually set because it's a small amount of data.

In [115]:
deposit_known = {
    'drei': 3,
    '3x': 3,
    '9150.-': 9150,
    '2xnkm': 2,
    '6000.-': 6000,
    '3-fache': 3,
    '3nkm': 3,    
}

deposit_unkown = ['ja', 'kaution', 'kaution:']

In [116]:
df['deposit'] = df['deposit'].apply(lambda value: str(value).lower())

In [117]:
df['deposit'] = df['deposit'].apply(lambda deposit: deposit_known.get(deposit, deposit))

### Fill in relative deposits
There are deposits that don't have the absolute amount defined. They are instead contain some small number (1,2,3). For such rows, the deposit is relative in terms of the monthly rent. 


In [118]:
def convert_to_number(value):
    if is_number(value):
        return float(value)
    else:
        return np.nan

converted_deposits = df['deposit'].apply(lambda x: convert_to_number(x))

# check which is the most common option for "relative" deposits 
converted_deposits[converted_deposits <= 5].value_counts()

3.0    144
2.0      4
2.5      1
Name: deposit, dtype: int64

Let's assume that the unclear values are three months, as it is the most common option.

In [119]:
# check if the value is in the list, if yes -> replace with 3, if not, leave it as is
def replace_unknown_deposit(old_deposit):
  if old_deposit in deposit_unkown:
    return 3
  else:
    return old_deposit 
 
df['deposit'] = df['deposit'].apply(replace_unknown_deposit)

# alternative with ternary operator https://book.pythontips.com/en/latest/ternary_operators.html
# value_if_true if condition else value_if_false
#df['deposit'] = df['deposit'].apply(lambda old_deposit: 3 if old_deposit in deposit_unkown else old_deposit)

In [120]:
df['deposit'].unique()

array(['3.0', '2920.53', 3, 'nan', '2400.0', '3507.42', '2000.0',
       '2139.78', '5397.00', '3885.00', '7950.00', '4152.0', '4650.00',
       '1500', 9150, '4545', 2, '6900.00', '5994.00', 6000, '7470.00',
       '5172.00', '2145.00', '6215.70', '11475.00', '14850.00', '4197.00',
       '2600.00', '12900.00', '2.0', '3450.0', '5000.0', '10000',
       '3620.0', '1690.0', '1790.0', '2990.0', '1490.0', '9960.00',
       '3159.00', '4350.00', '5010.0', '4333.90', '3000.0', '2013.00',
       '2841.0', '4000.0', '2850.72', '3300.75', '2344.0', '3645.00',
       '2397.0', '4500.0', '9718.00', '3597.00', '3900.0', '3780.00',
       '4800.0', '3582', '3630.0', '2549.79', '3831.0', '2274.00',
       '2850.0', '2960.0', '4680', '4350.0', '3500.0', '4080.0', '4050.0',
       '2338.0', '5940.00', '8100.0', '3859.17', '2630.43', '6552.0',
       '2700.00', '1797.00', '3345.0', '3360.0', '5148.0', '2472.0',
       '12000.00', '6300.00', '5394.00', '1554.00', '6600.0', '1727.76',
       '9025.50',

Let's convert all the non-na values to numbers

In [121]:
print(df['deposit'].dtype)

df['deposit'] = df['deposit'].astype(float) 

print(df['deposit'].dtype)

object
float64


The values 1, 2 and 3 refer to the number of months rent as deposit, so let's calculate those values by multiplying with the rent

In [122]:
print(df['rent'].dtype)
print(df['rent'].isna().sum())

float64
0


In [123]:
def convert_deposit_relative_to_absolute(row):
  deposit = row['deposit']
  if deposit <= 3:
      return row['rent'] * deposit
  else:
      return deposit

In [124]:
df['deposit'] = df.apply(convert_deposit_relative_to_absolute, axis = 1)

df[['security_deposit', 'deposit', 'rent']].head(20)

security_deposit   deposit     rent
0                 3   7977.00  2659.00
1           2920,53   2920.53  1200.00
2              Drei   2937.00   979.00
3               NaN       NaN  1830.22
4               NaN       NaN  2272.00
5                 3   2520.00   840.00
6               NaN       NaN  1509.45
7                 3   2192.19   730.73
8              Drei   1685.79   561.93
9                 3   5367.00  1789.00
10              NaN       NaN  2499.00
11              NaN       NaN  2328.00
12                3   7500.00  2500.00
13                3  19350.00  6450.00
14             2400   2400.00  1399.65
15                3   4516.80  1505.60
16          3507.42   3507.42  1429.10
17             2000   2000.00   930.00
18          2139.78   2139.78   868.06
19         5.397,00   5397.00  2179.00

### Check Distribution and Descriptive Stats

In [125]:
px.histogram(df, x="deposit", marginal='box')

In [126]:
df['deposit'].agg(['mean', 'median', 'std', 'min', 'max'])

mean       4670.031169
median     3900.000000
std        3153.089709
min         842.220000
max       29322.000000
Name: deposit, dtype: float64

### Filling in missing values (NaN) 

(Note: these are old advertisements and no longer available)

 - https://www.immobilienscout24.de/expose/116573177: has all of it's data mising, *will assume three months*
 - https://www.immobilienscout24.de/expose/115925878: has only some of its data missing, *will assume scraping error and therefore three months*
 - https://www.immobilienscout24.de/expose/114437800: has only deposit missing, *will assume there is no deposit*
 
**Because the data is missing for different reasons, we should use different fill strategies**

In [127]:
df[df['deposit'].isna()].head()

url  ... deposit
3   https://www.immobilienscout24.de/expose/116573177  ...     NaN
4   https://www.immobilienscout24.de/expose/115925878  ...     NaN
6   https://www.immobilienscout24.de/expose/108376992  ...     NaN
10  https://www.immobilienscout24.de/expose/114437800  ...     NaN
11  https://www.immobilienscout24.de/expose/116205341  ...     NaN

[5 rows x 22 columns]

In [128]:
standard_columns = ['url', 'region', 'condition', 'rooms', 'rent', 'year_construction', 'space', 'cold_rent', 'security_deposit']
optional_columns = list(set(df.columns) - set(standard_columns))

df['missing_count'] = df[optional_columns].isna().sum(axis=1)
df['missing_count']

0       5
1       5
2       4
3      13
4       8
       ..
759     5
760     8
761    13
762     4
763    13
Name: missing_count, Length: 764, dtype: int64

In [129]:
df[df['missing_count'] == 1].head(3)

url  ... missing_count
14  https://www.immobilienscout24.de/expose/116361020  ...             1
29  https://www.immobilienscout24.de/expose/116326551  ...             1
43  https://www.immobilienscout24.de/expose/116436290  ...             1

[3 rows x 23 columns]

In [130]:
def fill_missing_deposit(row):
  
    # if the value isn't missing, return the value
    if not pd.isna(row['deposit']):
        return row['deposit']
  
    # if there is only one missing value, return 0. Otherwise return 3 months of rent
    if row['missing_count'] == 1:
        return 0
    else:
        return row['rent'] * 3

In [131]:
df['deposit'] = df.apply(fill_missing_deposit, axis=1)

In [132]:
px.histogram(df, x="deposit", marginal='box')

In [133]:
df['deposit'].agg(['mean', 'median', 'std', 'min', 'max'])

mean       4854.710955
median     4197.000000
std        3295.911843
min         813.750000
max       42621.000000
Name: deposit, dtype: float64

In [134]:
df['deposit'].isna().sum()

0

## Clean `property_condition`

In [135]:
(df.isna().sum() / len(df)).sort_values(ascending = False)

parking_space          0.882199
renovated_date         0.744764
efficiency_class       0.671466
energy_requirement     0.573298
security_deposit       0.496073
cold_rent              0.442408
utilities              0.442408
heating_costs          0.442408
furnishing             0.417539
energy_certificate     0.409686
type_of_certificate    0.388743
energy_sources         0.353403
property_condition     0.344241
heating_type           0.325916
region                 0.000000
rent                   0.000000
condition              0.000000
rooms                  0.000000
missing_count          0.000000
year_construction      0.000000
space                  0.000000
deposit                0.000000
url                    0.000000
dtype: float64

In [136]:
df['property_condition'].value_counts()

New                     241
Mint Condition          103
Groomed                  65
Refurbished              29
After Renovation         26
Completely Renovated     21
Modernized               13
Needs Renovation          2
Well Maintained           1
Name: property_condition, dtype: int64

In [137]:
df['property_condition'].isna().sum()

263

In [138]:
df.head()

url  ... missing_count
0  https://www.immobilienscout24.de/expose/116051687  ...             5
1  https://www.immobilienscout24.de/expose/115338103  ...             5
2  https://www.immobilienscout24.de/expose/116458710  ...             4
3  https://www.immobilienscout24.de/expose/116573177  ...            13
4  https://www.immobilienscout24.de/expose/115925878  ...             8

[5 rows x 23 columns]

### Categorical Features need to be handled differently than Continuous

In [139]:
df['property_condition'].fillna('MISSING').value_counts()

MISSING                 263
New                     241
Mint Condition          103
Groomed                  65
Refurbished              29
After Renovation         26
Completely Renovated     21
Modernized               13
Needs Renovation          2
Well Maintained           1
Name: property_condition, dtype: int64

In [140]:
df['property_condition'] = df.property_condition.fillna('Unknown')

In [141]:
px.bar(df['property_condition'].value_counts().reset_index(), x='index', y='property_condition')

## Clean `cold_rent`

In [142]:
(df.isna().sum() / len(df)).sort_values(ascending=False)

parking_space          0.882199
renovated_date         0.744764
efficiency_class       0.671466
energy_requirement     0.573298
security_deposit       0.496073
cold_rent              0.442408
utilities              0.442408
heating_costs          0.442408
furnishing             0.417539
energy_certificate     0.409686
type_of_certificate    0.388743
energy_sources         0.353403
heating_type           0.325916
region                 0.000000
rent                   0.000000
condition              0.000000
rooms                  0.000000
missing_count          0.000000
year_construction      0.000000
space                  0.000000
deposit                0.000000
property_condition     0.000000
url                    0.000000
dtype: float64

**Since we have the warm rent for every sample, rather than use the median or mean, a better method might be to get the average warm/cold rent ratio and apply that to all missing values**

### Add NaN values

In [143]:
df['cold_rent'] = df['cold_rent'].replace('nan', np.nan)

### Compute cold to warm rent ratio

In [144]:
df['cold_rent'] = df['cold_rent'].apply(lambda x: str(x).replace('€', ''))
df['cold_rent'] = df['cold_rent'].apply(convert_decimal)

In [145]:
total_warm_and_cold = df[['rent', 'cold_rent']].dropna().astype(float).sum()
total_warm_and_cold

rent         1351180.56
cold_rent     672957.80
dtype: float64

In [146]:
warm_cold_rent_ratio = total_warm_and_cold['cold_rent'] / total_warm_and_cold['rent']
warm_cold_rent_ratio

0.4980517185652814

In [147]:
df['cold_rent'] = df['cold_rent'].astype(float)

In [148]:
px.histogram(df[['cold_rent']], x='cold_rent', marginal='box')

In [149]:
def get_cold_rent(row):
    if not pd.isna(row['cold_rent']): 
        return row['cold_rent']
    else:
        return row['rent'] * warm_cold_rent_ratio

In [150]:
df['cold_rent'] = df.apply(get_cold_rent, axis=1)

In [151]:
px.histogram(df, x='cold_rent', marginal='box')

*italicized text*## 0. Before further cleaning, print out how many missing values (sorted for readability) each column still has. Also check how many unique values there are for each column

In [152]:

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 764 entries, 0 to 763
Data columns (total 23 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   url                  764 non-null    object 
 1   region               764 non-null    object 
 2   condition            764 non-null    object 
 3   rooms                764 non-null    float64
 4   rent                 764 non-null    float64
 5   year_construction    764 non-null    int64  
 6   space                764 non-null    float64
 7   cold_rent            764 non-null    float64
 8   utilities            426 non-null    object 
 9   heating_costs        426 non-null    object 
 10  security_deposit     385 non-null    object 
 11  parking_space        90 non-null     object 
 12  renovated_date       195 non-null    float64
 13  property_condition   764 non-null    object 
 14  furnishing           445 non-null    object 
 15  heating_type         515 non-null    obj

In [153]:
# to get the count of null/NaN values for each column 
df.isna().sum().sort_values(ascending=False)

parking_space          674
renovated_date         569
efficiency_class       513
energy_requirement     438
security_deposit       379
utilities              338
heating_costs          338
furnishing             319
energy_certificate     313
type_of_certificate    297
energy_sources         270
heating_type           249
year_construction        0
region                   0
condition                0
rooms                    0
rent                     0
missing_count            0
space                    0
cold_rent                0
deposit                  0
property_condition       0
url                      0
dtype: int64

In [154]:
# get the number of unique values for each column
df.nunique().sort_values()

energy_certificate       3
heating_type             7
efficiency_class         8
furnishing               8
condition                9
region                  10
property_condition      10
rooms                   11
renovated_date          12
missing_count           13
parking_space           21
energy_sources          29
type_of_certificate     85
year_construction       93
heating_costs          111
energy_requirement     113
security_deposit       216
utilities              261
space                  571
deposit                626
rent                   639
cold_rent              650
url                    764
dtype: int64

## 1. Clean: `heating_type`


In [155]:
# From the previous cell, it's noticable that there are very few distinct heating types 
# (only 7 unique values). That means we're dealing with a discrete/categorical value here. 
df['heating_type'].unique()

array(['Underfloor Heating', 'Central Heating', nan, 'District Heating',
       'Floor Heating', 'Gas Heating', 'CHP Plants',
       'Wood Pellet Heating'], dtype=object)

In [156]:
# The 'update_xaxes' is just there to sort the bars
px.histogram(df, x='heating_type').update_xaxes(categoryorder="total descending")

In [157]:
# Notice that the heating_type actually has 249 NaN values. The plotting doesn't reflect this by default.
# You can fill in all NaN values with some custom default value to make it visible in the plot.
df['heating_type'] = df['heating_type'].fillna('UNKOWN')
px.histogram(df, x='heating_type').update_xaxes(categoryorder="total descending")

## 2. Clean: `renovated_date`



In [158]:
# renovated_date has a lot of missing values. The existing renovations are concentrated
# after 2015 as the plot shows. 
px.histogram(df, x='renovated_date')

In [159]:
# From the .info() call you see that there is no mix between numeric and string values.
# So each date represents a correct numeric value. 
# Here it doesn't make much sense to replace NaN values with any sort of hardcoded 
# custom value. Depending on the context, NaN might mean that the appartment was never
# renovated. 
# For an exercise we could simply replace missing values with the corresponding 
# construction year. We know that the construction year is also a valid numeric column.
# Additionally, it does not have any missing values. 

# The 'fillna' function is flexible enough and can also take a series of values
# that will be used if the value in the target column (this case 'renovated_date') 
# is missing. 
df['renovated_date_filled'] = df['renovated_date'].fillna(df['year_construction'])

# To make sure that the missing values are replaced with the construction year,
# but the non-missing values stay the same
df[['year_construction', 'renovated_date', 'renovated_date_filled']].head(30)

year_construction  renovated_date  renovated_date_filled
0                2019             NaN                 2019.0
1                2020             NaN                 2020.0
2                1997             NaN                 1997.0
3                1900             NaN                 1900.0
4                2020             NaN                 2020.0
5                1997          2019.0                 2019.0
6                2015             NaN                 2015.0
7                1900             NaN                 1900.0
8                1971             NaN                 1971.0
9                2020             NaN                 2020.0
10               2019             NaN                 2019.0
11               2020             NaN                 2020.0
12               2019             NaN                 2019.0
13               2016             NaN                 2016.0
14               2019          2019.0                 2019.0
15               2018             NaN                 2018.0
16               2020             NaN                 2020.0
17               2018             NaN                 2018.0
18               2020             NaN                 2020.0
19               2019             NaN                 2019.0
20               2019             NaN                 2019.0
21               1994          2019.0                 2019.0
22               2019             NaN                 2019.0
23               2020             NaN                 2020.0
24               1918          2019.0                 2019.0
25               2020             NaN                 2020.0
26               2014             NaN                 2014.0
27               2015             NaN                 2015.0
28               2019             NaN                 2019.0
29               2016             NaN                 2016.0

In [160]:
# A more complex approach could be to apply some sort of interpolation 
# based on the construction year. A simple initial check could be to compute
# the difference between the renovation and construction (only for existing renovated_date values).
# Based on that you could derive a function that takes the construction year as input and 
# returns the estimated renovated_date as output

# Again, if this makes sense or not depends on the context (maybe a missing value just means there wasn't a renovation at all)

# get the construction year and renovated date columns only for rows where we have an existing renovated_date value
dates = df.loc[~df['renovated_date'].isna(), ['year_construction', 'renovated_date']] 
# compute the difference and check the results
dates['diff'] = dates['renovated_date'] - dates['year_construction']
dates.head(20)

year_construction  renovated_date   diff
5                 1997          2019.0   22.0
14                2019          2019.0    0.0
21                1994          2019.0   25.0
24                1918          2019.0  101.0
43                2016          2016.0    0.0
51                1918          2019.0  101.0
53                1905          2016.0  111.0
60                1918          2019.0  101.0
70                1902          2020.0  118.0
71                1900          2020.0  120.0
72                1910          2018.0  108.0
78                2015          2015.0    0.0
94                2020          2020.0    0.0
109               1998          2017.0   19.0
112               1904          2019.0  115.0
116               1900          2019.0  119.0
121               1919          2018.0   99.0
124               1959          2019.0   60.0
126               1887          2019.0  132.0
128               2018          2020.0    2.0

## 3. Clean: `parking_space`


In [161]:
# The parking_space column clearly shows the costs. However, the values are strings,
# many having a € suffix. Additionally, they are using the german formatting which
# makes it difficult to parse the values directly. 
# To get the values into a workable, they'll have to be transformed by removing 
# the €-symbol and correcting the formatting. 
df['parking_space'].unique()

array(['120€', nan, '150€', '200€', '0€', '130€', '90€', '350€', '119€',
       '100€', '60€', '140€', '220€', '70€', '110€', '145€', '166,6€',
       '180€', '250€', '149€', '280€', '75€'], dtype=object)

In [162]:
# helper function from the lecture to adjust the formatting 
def convert_decimal(value):
    """Convert numeric string to a decimal number."""

    def convert_string_to_numeric_representation(value):
      """Replace a period with empty stringm and comma with period (decimal point)."""
      return value.replace(".", "").replace(",", ".")
    
    # return NaN if missing
    if pd.isna(value):
        return np.nan       
    
    # check if wrong format without decimal i.e. 3.500
    if len(value.split('.')[-1]) == 3:
        return convert_string_to_numeric_representation(value)

    # if it's convertable to float, just return the value
    # otherwise correct the formatting
    try:                               
        float(value)
        return value
    except:
        return convert_string_to_numeric_representation(value)

In [163]:
# strip removes leading and trailing whitespace characters
with_removed_symbol = df['parking_space'].str.replace('€', '').str.strip() 

with_removed_symbol.unique()

array(['120', nan, '150', '200', '0', '130', '90', '350', '119', '100',
       '60', '140', '220', '70', '110', '145', '166,6', '180', '250',
       '149', '280', '75'], dtype=object)

In [164]:
# Transform the formatting
# After this step all values can easily be cast to their numeric representation 
cleaned_parking_space = with_removed_symbol.apply(convert_decimal)
cleaned_parking_space.unique()

array(['120', nan, '150', '200', '0', '130', '90', '350', '119', '100',
       '60', '140', '220', '70', '110', '145', '166.6', '180', '250',
       '149', '280', '75'], dtype=object)

In [165]:
cleaned_parking_space = cleaned_parking_space.astype(float)

# Note the type of the columns 'dtype: float64'. This means all values are of type float.
cleaned_parking_space

0      120.0
1        NaN
2        NaN
3        NaN
4        NaN
       ...  
759      NaN
760      NaN
761      NaN
762      NaN
763      NaN
Name: parking_space, Length: 764, dtype: float64

In [166]:
# Note that missing NaN values are also interpreted as float by pandas. That does not mean
# you can do any "numeric computations" on it.
type(np.nan)

float

In [167]:
# Different approaches can be taken to fill in the missing values. 
# But note that this column is very sparse, meaning there are far more
# missing than existing values. So a more real-world approach should also take this into account.
print(cleaned_parking_space.isna().sum())
print((~cleaned_parking_space.isna()).sum())

674
90


In [168]:
# This shows some summary statistics based on the existing parking_space values. 
cleaned_parking_space.describe()

count     90.000000
mean     139.662222
std       71.774795
min        0.000000
25%      100.000000
50%      120.000000
75%      150.000000
max      350.000000
Name: parking_space, dtype: float64

In [169]:
# Replace missing parking_space values with the median (50% above) 
df['parking_space_cleaned'] = cleaned_parking_space.fillna(cleaned_parking_space.median())

# This is one approach, but notice how this completely distorts the values. 
# We're filling in a derived value from a very small set (90 existing values overall)
# and using it to fill in the remaining 674 values. This is probably too simplistic
# and in a real-world scenario it would be interesting to explore more complex approaches further
# For example, maybe there is some relation between other parameters (rent, size of apartment, area)
# and the cost for parking. Based on that, a more realistic replacement could be applied.

# See the next section about 'utilities' for a different approach. 
px.histogram(df, x='parking_space_cleaned')

## 4. Clean: `utilities`



In [170]:
# Similar cleaning as for the parking space
# Symbols like + and € have to be removed and the formatting needs to be adjusted
# before the existing values can be correctly parsed to numbers
df['utilities'].head(20)

0        +460€
1     +226,49€
2         +90€
3          NaN
4          NaN
5        +110€
6          NaN
7        +110€
8        +110€
9        +255€
10       +500€
11         NaN
12       +250€
13       +750€
14    +133,30€
15    +225,84€
16    +164,09€
17       +170€
18     +98,38€
19       +380€
Name: utilities, dtype: object

In [171]:
# remove the irrelevant symbols from the values
utilities_cleaned = df['utilities'].str.replace('+', '').str.replace('€', '').str.strip()

utilities_cleaned.head(20)

0        460
1     226,49
2         90
3        NaN
4        NaN
5        110
6        NaN
7        110
8        110
9        255
10       500
11       NaN
12       250
13       750
14    133,30
15    225,84
16    164,09
17       170
18     98,38
19       380
Name: utilities, dtype: object

In [172]:
# reformat each number to a parseable format and replace values of 'keine' (= none in german)
# with 0
utilities_cleaned = utilities_cleaned.apply(convert_decimal)

utilities_cleaned = utilities_cleaned.apply(lambda x: '0' if x == 'keine' else x)

utilities_cleaned.head(20)

0        460
1     226.49
2         90
3        NaN
4        NaN
5        110
6        NaN
7        110
8        110
9        255
10       500
11       NaN
12       250
13       750
14    133.30
15    225.84
16    164.09
17       170
18     98.38
19       380
Name: utilities, dtype: object

In [181]:
# cast all values to numeric types
utilities_cleaned = utilities_cleaned.astype(float)

## 5. Clean: `energy_requirement`


In [174]:
print(f"Missing  values: {df['energy_requirement'].isna().sum()}")
print(f"Existing values: {(~df['energy_requirement'].isna()).sum()}")

df['energy_requirement'].sample(20)

Missing  values: 438
Existing values: 326


37     76,9 kWh/(m²*a)
303                NaN
165                NaN
368                NaN
327                NaN
586                NaN
414                NaN
286                NaN
562                NaN
448    66,2 kWh/(m²*a)
310      59 kWh/(m²*a)
284                NaN
346      72 kWh/(m²*a)
32                 NaN
699                NaN
531                NaN
480                NaN
737                NaN
246                NaN
207     170 kWh/(m²*a)
Name: energy_requirement, dtype: object

In [175]:
# Similar to the previous examples, there are some symbols and formatting
# that have to be adjusted to make the existing values parseable to numbers.
# Note the regex=False flag. By default, the replace function treats the target
# string that should be replaced as a regex https://en.wikipedia.org/wiki/Regular_expression.
# Because the string that we want to replace contains some special regex-specific characters,
# the replacement doesn't work without turning off the flag.
energy_req = df['energy_requirement'].str.replace('kWh/(m²*a)', '', regex=False).str.strip()

energy_req.sample(10)

344      NaN
675      NaN
43        74
725      NaN
169      NaN
331    155,5
261      NaN
575     66,2
48        68
121    128,1
Name: energy_requirement, dtype: object

In [176]:
# The other steps are just as before. Reformatting and casting to float. 
energy_req_cleaned = energy_req.apply(convert_decimal)
energy_req_cleaned = energy_req_cleaned.astype(float)
energy_req_cleaned.sample(10)

117     NaN
288    66.0
335     NaN
328     NaN
412     NaN
157     NaN
562     NaN
690     NaN
452    42.8
174     NaN
Name: energy_requirement, dtype: float64

In [177]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 764 entries, 0 to 763
Data columns (total 25 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   url                    764 non-null    object 
 1   region                 764 non-null    object 
 2   condition              764 non-null    object 
 3   rooms                  764 non-null    float64
 4   rent                   764 non-null    float64
 5   year_construction      764 non-null    int64  
 6   space                  764 non-null    float64
 7   cold_rent              764 non-null    float64
 8   utilities              426 non-null    object 
 9   heating_costs          426 non-null    object 
 10  security_deposit       385 non-null    object 
 11  parking_space          90 non-null     object 
 12  renovated_date         195 non-null    float64
 13  property_condition     764 non-null    object 
 14  furnishing             445 non-null    object 
 15  heatin

In [178]:
df['energy_req_cleaned'] = energy_req_cleaned

# In this case the relationship is not that clear (you can try it out with different columns except rent)
# So this would need a more complex approach to fill in the missing values
px.scatter(df, x='rent', y='energy_req_cleaned')

##  Clean: `type_of_certificate`



In [179]:
# We see that most values also contain the concrete amount for the energy requirement (basically what is in the energy_requirement column already).

# It's possible to extract the numeric amount from the string and create a 
# discrete categorical value that just describes the type of certificate. 
df['type_of_certificate'].unique()

array(['Bedarfsausweis', nan,
       'Verbrauchsausweis Energie\xadver\xadbrauchs\xadkennwert 98,5 kWh/(m²*a)',
       'Verbrauchsausweis Energie\xadver\xadbrauchs\xadkennwert 84 kWh/(m²*a)',
       'Verbrauchsausweis End\xadenergie\xadverbrauch 171,7 kWh/(m²*a)',
       'Verbrauchsausweis Energie\xadver\xadbrauchs\xadkennwert 74 kWh/(m²*a) Energieverbrauch für Warmwasser enthalten',
       'Verbrauchsausweis End\xadenergie\xadverbrauch 109 kWh/(m²*a)',
       'Verbrauchsausweis Energie\xadver\xadbrauchs\xadkennwert 120,8 kWh/(m²*a)',
       'Verbrauchsausweis End\xadenergie\xadverbrauch 204,3 kWh/(m²*a)',
       'Verbrauchsausweis Energie\xadver\xadbrauchs\xadkennwert 215 kWh/(m²*a)',
       'Verbrauchsausweis End\xadenergie\xadverbrauch 218,2 kWh/(m²*a)',
       'Verbrauchsausweis End\xadenergie\xadverbrauch 120 kWh/(m²*a)',
       'Verbrauchsausweis',
       'Verbrauchsausweis End\xadenergie\xadverbrauch 215 kWh/(m²*a)',
       'Verbrauchsausweis End\xadenergie\xadverbrauch 219,5 kW

In [180]:
# First, try to extract the digits from the text.
# The provided regex [0-9]+ matches 1 or more consecutive digits.
# 'extract' only returns the first match of the regex pattern. 
# For example, for a value 172,5 this would only extract 172 and not 5
# (because there is a non-digit ',' and 172 is the first match of consecutive digits). 
# In this case this is good enough, since we don't have values that go beyond
# hundreds. To make it more robust, it would be good to keep all matches of the regex ('extractall' can do that, but the output is than a bit more complex)
# and/or have a more complex regex pattern 
extracted_numbers = df['type_of_certificate'].str.extract('([0-9]+)')

# The extracted numbers can for example be used to fill in some missing values
# for the "energy_requirements" columns. So if the energy_requirement value is missing
# but the corresponding extracted number from here exists, that can be used. 

# Just to show a couple of the extracted values
extracted_numbers.astype(float).dropna().sample(10)

0
334   72.0
109  219.0
569   64.0
171   98.0
232  121.0
289  105.0
192   78.0
509  126.0
134  126.0
124  163.0